File: packages/browseros-agent/packages/browser-core/src/core/snapshot/diff.ts
const BULLET = /^(\s*)- /
const MAX_LCS_CELLS = 4_000_000

export interface SnapshotDiff {
  /** Line-level diff or bounded change summary; empty when nothing changed. */
  text: string
  added: number
  removed: number
  changed: boolean
  /** Counts are changed-window sizes because the line-level comparison was skipped. */
  lineDiffSkipped?: true
  urlChanged?: true
  beforeUrl?: string
  afterUrl?: string
}

export interface DiffOptions {
  contextRadius?: number
}

export interface SnapshotObservation {
  text: string
  url?: string
}

interface TaggedLine {
  gutter: ' ' | '-' | '+'
  text: string
}

interface ChangedWindow {
  start: number
  beforeEnd: number
  afterEnd: number
}

/**
 * Computes a compact line diff between rendered snapshots, or a bounded summary when the changed
 * window is too large for LCS. Distant unchanged context is elided from line-level results.
 */
export function diffSnapshots(
  before: string,
  after: string,
  opts: DiffOptions = {},
): SnapshotDiff {
  if (before === after) {
    return { text: '', added: 0, removed: 0, changed: false }
  }

  const beforeLines = splitLines(before)
  const afterLines = splitLines(after)
  const { start, beforeEnd, afterEnd } = findChangedWindow(
    beforeLines,
    afterLines,
  )
  const removedWindowLines = beforeEnd - start
  const addedWindowLines = afterEnd - start

  if (exceedsLcsBudget(removedWindowLines, addedWindowLines)) {
    return {
      text: [
        `Snapshot changed substantially: ${beforeLines.length} lines before, ${afterLines.length} lines after.`,
        `Line-level diff skipped because the changed region exceeds the ${MAX_LCS_CELLS}-cell comparison limit.`,
      ].join('\n'),
      added: addedWindowLines,
      removed: removedWindowLines,
      changed: true,
      lineDiffSkipped: true,
    }
  }

  const tagged: TaggedLine[] = []
  for (let i = 0; i < start; i++) {
    tagged.push({ gutter: ' ', text: beforeLines[i] })
  }
  appendDiffLines(
    tagged,
    beforeLines.slice(start, beforeEnd),
    afterLines.slice(start, afterEnd),
  )
  for (let i = beforeEnd; i < beforeLines.length; i++) {
    tagged.push({ gutter: ' ', text: beforeLines[i] })
  }

  let added = 0
  let removed = 0
  for (const line of tagged) {
    if (line.gutter === '+') added++
    else if (line.gutter === '-') removed++
  }

  const body = collapse(tagged, opts.contextRadius ?? 3)
  return {
    text: `${body}\n${added} added, ${removed} removed`,
    added,
    removed,
    changed: true,
  }
}

/** Compares successive page observations, returning the full snapshot when navigation changed the URL. */
export function diffSnapshotObservations(
  before: SnapshotObservation | undefined,
  after: SnapshotObservation,
  opts: DiffOptions = {},
): SnapshotDiff {
  const beforeUrl = before?.url
  const afterUrl = after.url
  if (isKnownUrl(beforeUrl) && isKnownUrl(afterUrl) && beforeUrl !== afterUrl) {
    return {
      text: after.text,
      added: 0,
      removed: 0,
      changed: true,
      urlChanged: true,
      beforeUrl,
      afterUrl,
    }
  }

  const diff = diffSnapshots(before?.text ?? '', after.text, opts)
  if (isKnownUrl(afterUrl)) return { ...diff, afterUrl }
  return diff
}

function isKnownUrl(url: string | undefined): url is string {
  return url !== undefined && url !== '' && url !== 'unknown'
}

function splitLines(value: string): string[] {
  return value === '' ? [] : value.split('\n')
}

function findChangedWindow(before: string[], after: string[]): ChangedWindow {
  let start = 0
  while (
    start < before.length &&
    start < after.length &&
    before[start] === after[start]
  ) {
    start++
  }

  let beforeEnd = before.length
  let afterEnd = after.length
  while (
    beforeEnd > start &&
    afterEnd > start &&
    before[beforeEnd - 1] === after[afterEnd - 1]
  ) {
    beforeEnd--
    afterEnd--
  }

  return { start, beforeEnd, afterEnd }
}

function exceedsLcsBudget(beforeLength: number, afterLength: number): boolean {
  return (
    beforeLength > 0 &&
    afterLength > 0 &&
    beforeLength + 1 > MAX_LCS_CELLS / (afterLength + 1)
  )
}

function appendDiffLines(
  tagged: TaggedLine[],
  before: string[],
  after: string[],
): void {
  if (before.length === 0) {
    for (const text of after) tagged.push({ gutter: '+', text })
    return
  }
  if (after.length === 0) {
    for (const text of before) tagged.push({ gutter: '-', text })
    return
  }

  const table = buildLcsTable(before, after)
  let i = 0
  let j = 0

  while (i < before.length && j < after.length) {
    if (before[i] === after[j]) {
      tagged.push({ gutter: ' ', text: before[i] })
      i++
      j++
    } else if (table[i + 1][j] >= table[i][j + 1]) {
      tagged.push({ gutter: '-', text: before[i] })
      i++
    } else {
      tagged.push({ gutter: '+', text: after[j] })
      j++
    }
  }

  while (i < before.length) {
    tagged.push({ gutter: '-', text: before[i++] })
  }
  while (j < after.length) {
    tagged.push({ gutter: '+', text: after[j++] })
  }
}

function buildLcsTable(before: string[], after: string[]): number[][] {
  const table = Array.from({ length: before.length + 1 }, () =>
    new Array<number>(after.length + 1).fill(0),
  )

  for (let i = before.length - 1; i >= 0; i--) {
    for (let j = after.length - 1; j >= 0; j--) {
      table[i][j] =
        before[i] === after[j]
          ? table[i + 1][j + 1] + 1
          : Math.max(table[i + 1][j], table[i][j + 1])
    }
  }

  return table
}

/** Keep changed lines plus a `radius` window of context; elide gaps with `…`. */
function collapse(tagged: TaggedLine[], radius: number): string {
  const keep = new Array<boolean>(tagged.length).fill(false)
  for (let i = 0; i < tagged.length; i++) {
    if (tagged[i].gutter === ' ') continue
    const lo = Math.max(0, i - radius)
    const hi = Math.min(tagged.length - 1, i + radius)
    for (let j = lo; j <= hi; j++) keep[j] = true
  }

  const out: string[] = []
  let prev = -1
  for (let i = 0; i < tagged.length; i++) {
    if (!keep[i]) continue
    if (prev >= 0 && i - prev > 1) out.push('…')
    const { gutter, text } = tagged[i]
    out.push(`${gutter} ${text.replace(BULLET, '$1')}`)
    prev = i
  }
  return out.join('\n')
}

File: packages/browseros-agent/crates/browseros-core/src/snapshot/diff.rs
#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct SnapshotDiff {
    pub text: String,
    pub added: usize,
    pub removed: usize,
    pub changed: bool,
    /// When true, `text` is the full current snapshot for a known URL change, not a line diff,
    /// and `added` and `removed` are zero.
    pub url_changed: bool,
    pub before_url: Option<String>,
    pub after_url: Option<String>,
}

#[derive(Debug, Clone, Copy, Default, PartialEq, Eq)]
pub struct DiffOptions {
    pub context_radius: Option<usize>,
}

#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct SnapshotObservation {
    pub text: String,
    pub url: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
struct TaggedLine {
    gutter: char,
    text: String,
}

pub fn diff_snapshots(before: &str, after: &str, opts: DiffOptions) -> SnapshotDiff {
    if before == after {
        return SnapshotDiff::default();
    }

    let mut tagged = Vec::new();
    let mut added = 0;
    let mut removed = 0;
    for line in diff_lines(before, after) {
        if line.gutter == '+' {
            added += 1;
        } else if line.gutter == '-' {
            removed += 1;
        }
        tagged.push(line);
    }

    let body = collapse(&tagged, opts.context_radius.unwrap_or(3));
    SnapshotDiff {
        text: format!("{body}\n{added} added, {removed} removed"),
        added,
        removed,
        changed: true,
        url_changed: false,
        before_url: None,
        after_url: None,
    }
}

pub fn diff_snapshot_observations(
    before: Option<&SnapshotObservation>,
    after: &SnapshotObservation,
    opts: DiffOptions,
) -> SnapshotDiff {
    let before_url = before.and_then(|before| before.url.as_deref());
    let after_url = after.url.as_deref();
    if is_known_url(before_url) && is_known_url(after_url) && before_url != after_url {
        return SnapshotDiff {
            text: after.text.clone(),
            added: 0,
            removed: 0,
            changed: true,
            url_changed: true,
            before_url: before_url.map(ToString::to_string),
            after_url: after_url.map(ToString::to_string),
        };
    }

    let before_text = before.map(|before| before.text.as_str()).unwrap_or("");
    let mut diff = diff_snapshots(before_text, &after.text, opts);
    if is_known_url(after_url) {
        diff.after_url = after_url.map(ToString::to_string);
    }
    diff
}

fn is_known_url(url: Option<&str>) -> bool {
    url.is_some_and(|url| !url.is_empty() && url != "unknown")
}

fn split_lines(value: &str) -> Vec<&str> {
    if value.is_empty() {
        Vec::new()
    } else {
        value.split('\n').collect()
    }
}

fn diff_lines(before: &str, after: &str) -> Vec<TaggedLine> {
    let before_lines = split_lines(before);
    let after_lines = split_lines(after);
    let table = build_lcs_table(&before_lines, &after_lines);
    let mut tagged = Vec::new();
    let mut i = 0;
    let mut j = 0;

    while i < before_lines.len() && j < after_lines.len() {
        if before_lines[i] == after_lines[j] {
            tagged.push(TaggedLine {
                gutter: ' ',
                text: before_lines[i].to_string(),
            });
            i += 1;
            j += 1;
        } else if table[i + 1][j] >= table[i][j + 1] {
            tagged.push(TaggedLine {
                gutter: '-',
                text: before_lines[i].to_string(),
            });
            i += 1;
        } else {
            tagged.push(TaggedLine {
                gutter: '+',
                text: after_lines[j].to_string(),
            });
            j += 1;
        }
    }

    while i < before_lines.len() {
        tagged.push(TaggedLine {
            gutter: '-',
            text: before_lines[i].to_string(),
        });
        i += 1;
    }
    while j < after_lines.len() {
        tagged.push(TaggedLine {
            gutter: '+',
            text: after_lines[j].to_string(),
        });
        j += 1;
    }

    tagged
}

fn build_lcs_table(before: &[&str], after: &[&str]) -> Vec<Vec<usize>> {
    let mut table = vec![vec![0; after.len() + 1]; before.len() + 1];
    for i in (0..before.len()).rev() {
        for j in (0..after.len()).rev() {
            table[i][j] = if before[i] == after[j] {
                table[i + 1][j + 1] + 1
            } else {
                table[i + 1][j].max(table[i][j + 1])
            };
        }
    }
    table
}

fn collapse(tagged: &[TaggedLine], radius: usize) -> String {
    let mut keep = vec![false; tagged.len()];
    for (index, line) in tagged.iter().enumerate() {
        if line.gutter == ' ' {
            continue;
        }
        let lo = index.saturating_sub(radius);
        let hi = (index + radius).min(tagged.len().saturating_sub(1));
        for item in keep.iter_mut().take(hi + 1).skip(lo) {
            *item = true;
        }
    }

    let mut out = Vec::new();
    let mut prev = None;
    for (index, line) in tagged.iter().enumerate() {
        if !keep[index] {
            continue;
        }
        if prev.is_some_and(|prev| index - prev > 1) {
            out.push("…".to_string());
        }
        out.push(format!("{} {}", line.gutter, strip_bullet(&line.text)));
        prev = Some(index);
    }
    out.join("\n")
}

fn strip_bullet(line: &str) -> String {
    let trimmed = line.trim_start();
    if let Some(rest) = trimmed.strip_prefix("- ") {
        let leading = line.len() - trimmed.len();
        format!("{}{}", &line[..leading], rest)
    } else {
        line.to_string()
    }
}

#[cfg(test)]
mod tests {
    use super::{
        DiffOptions, SnapshotDiff, SnapshotObservation, diff_snapshot_observations, diff_snapshots,
    };

    #[test]
    fn identical_snapshots_short_circuit_to_no_change() {
        let snap = "- button \"Go\" [ref=e1]";
        assert_eq!(
            diff_snapshots(snap, snap, DiffOptions::default()),
            SnapshotDiff::default()
        );
    }

    #[test]
    fn state_change_shows_removed_added_pair_on_same_ref() {
        let before = "- button \"Save\" [ref=e1]";
        let after = "- button \"Save\" [ref=e1] [disabled]";
        let diff = diff_snapshots(before, after, DiffOptions::default());
        assert!(diff.changed);
        assert_eq!(diff.added, 1);
        assert_eq!(diff.removed, 1);
        assert!(diff.text.contains("- button \"Save\" [ref=e1]"));
        assert!(diff.text.contains("+ button \"Save\" [ref=e1] [disabled]"));
        assert!(diff.text.contains("1 added, 1 removed"));
    }

    #[test]
    fn pure_additions_count_only_as_added_and_strip_bullet() {
        let before = "- main\n  - link \"Home\" [ref=e1]";
        let after = "- main\n  - link \"Home\" [ref=e1]\n  - link \"About\" [ref=e2]";
        let diff = diff_snapshots(before, after, DiffOptions::default());
        assert_eq!(diff.added, 1);
        assert_eq!(diff.removed, 0);
        assert!(diff.text.contains("+   link \"About\" [ref=e2]"));
    }

    #[test]
    fn stable_refs_turn_top_insertions_into_one_added_line() {
        let before = "- button \"A\" [ref=e1]\n- link \"B\" [ref=e2]";
        let after = [
            "- button \"X\" [ref=e3]",
            "- button \"A\" [ref=e1]",
            "- link \"B\" [ref=e2]",
        ]
        .join("\n");
        let diff = diff_snapshots(before, &after, DiffOptions::default());
        assert_eq!(diff.added, 1);
        assert_eq!(diff.removed, 0);
        assert!(diff.text.contains("+ button \"X\" [ref=e3]"));
        assert!(diff.text.contains("1 added, 0 removed"));
    }

    #[test]
    fn collapses_far_apart_context_with_ellipsis() {
        let before = (0..30)
            .map(|index| format!("- item {index}"))
            .collect::<Vec<_>>()
            .join("\n");
        let after = before
            .replace("- item 0", "- item ZERO")
            .replace("- item 29", "- item LAST");
        let diff = diff_snapshots(
            &before,
            &after,
            DiffOptions {
                context_radius: Some(2),
            },
        );
        assert!(diff.text.contains('…'));
        assert!(diff.text.contains("- item 0"));
        assert!(diff.text.contains("+ item ZERO"));
        assert!(diff.text.contains("+ item LAST"));
        assert!(!diff.text.contains("item 15"));
    }

    #[test]
    fn url_changes_return_full_current_snapshot() {
        let before = SnapshotObservation {
            text: "- main\n  - button \"Old page\" [ref=e1]".to_string(),
            url: Some("https://example.com/old".to_string()),
        };
        let after = SnapshotObservation {
            text: "- main\n  - heading \"New page\"".to_string(),
            url: Some("https://example.com/new".to_string()),
        };
        let diff = diff_snapshot_observations(Some(&before), &after, DiffOptions::default());
        assert_eq!(diff.text, after.text);
        assert_eq!(diff.added, 0);
        assert_eq!(diff.removed, 0);
        assert!(diff.changed);
        assert!(diff.url_changed);
        assert_eq!(diff.before_url.as_deref(), before.url.as_deref());
        assert_eq!(diff.after_url.as_deref(), after.url.as_deref());
    }

    #[test]
    fn unknown_urls_keep_line_diff_behavior() {
        let diff = diff_snapshot_observations(
            Some(&SnapshotObservation {
                text: "- main\n  - button \"Old\"".to_string(),
                url: Some("unknown".to_string()),
            }),
            &SnapshotObservation {
                text: "- main\n  - button \"New\"".to_string(),
                url: Some("https://example.com/new".to_string()),
            },
            DiffOptions::default(),
        );
        assert!(diff.changed);
        assert!(!diff.url_changed);
        assert_eq!(diff.added, 1);
        assert_eq!(diff.removed, 1);
        assert!(diff.text.contains("-   button \"Old\""));
        assert!(diff.text.contains("+   button \"New\""));
    }

    #[test]
    fn same_url_diffs_preserve_current_url() {
        let diff = diff_snapshot_observations(
            Some(&SnapshotObservation {
                text: "- main\n  - button \"Save\" [ref=e1]".to_string(),
                url: Some("https://example.com/form".to_string()),
            }),
            &SnapshotObservation {
                text: "- main\n  - button \"Saved\" [ref=e1]".to_string(),
                url: Some("https://example.com/form".to_string()),
            },
            DiffOptions::default(),
        );
        assert!(diff.changed);
        assert!(!diff.url_changed);
        assert_eq!(diff.after_url.as_deref(), Some("https://example.com/form"));
    }
}
